ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/LogSegment.java

```diff
@@ -246,13 +247,48 @@ public LogEntryProto load(LogRecord key) throws IOException {
     }
   }
 
+  static class EntryCache {
+    private final Map<TermIndex, ReferenceCountedObject<LogEntryProto>> map = new ConcurrentHashMap<>();
+    private final AtomicLong size = new AtomicLong();
+
+    long size() {
+      return size.get();
+    }
+
+    LogEntryProto get(TermIndex ti) {
+      return Optional.ofNullable(map.get(ti))
+          .map(ReferenceCountedObject::get)
+          .orElse(null);
+    }
+
+    void clear() {
+      map.values().forEach(ReferenceCountedObject::release);
+      map.clear();
+      size.set(0);
+    }
+
+    void put(TermIndex key, ReferenceCountedObject<LogEntryProto> valueRef, Op op) {
+      valueRef.retain();
+      Optional.ofNullable(map.put(key, valueRef)).ifPresent(this::release);
+      size.getAndAdd(getEntrySize(valueRef.get(), op));
+    }
+
+    private void release(ReferenceCountedObject<LogEntryProto> entry) {
+      size.getAndAdd(-getEntrySize(entry.get(), Op.REMOVE_CACHE));
+      entry.release();
```

Comment:
`getEntrySize` may result in different results based on `op`. If `put` and `release` use different `op`, may the `size` calculation be wrong?

Reply:
You are right that `getEntrySize(..)` seems returning different sizes for different op, although the size returned are always the same in the current usage.  Let me refactor it.